src/mlpack/methods/amf/init_rules/given_init.hpp

```diff
@@ -28,32 +28,89 @@ class GivenInitialization
 {
  public:
   // Empty constructor required for the InitializeRule template.
-  GivenInitialization() { }
+  GivenInitialization() : wIsGiven(false), hIsGiven(false) { }
 
   // Initialize the GivenInitialization object with the given matrices.
-  GivenInitialization(const arma::mat& w, const arma::mat& h) : w(w), h(h) { }
+  GivenInitialization(const arma::mat& w, const arma::mat& h) :
+    w(w), h(h), wIsGiven(true), hIsGiven(true) { }
 
   // Initialize the GivenInitialization object, taking control of the given
   // matrices.
   GivenInitialization(const arma::mat&& w, const arma::mat&& h) :
     w(std::move(w)),
-    h(std::move(h))
+    h(std::move(h)),
+    wIsGiven(true),
+    hIsGiven(true)
   { }
 
+  // Initialize either H or W with the given matrix.
+  GivenInitialization(const char whichMatrix, const arma::mat& m)
+  {
+    if (whichMatrix == 'W' || whichMatrix == 'w')
+    {
+      w = m;
+      wIsGiven = true;
+      hIsGiven = false;
+    }
+    else if (whichMatrix == 'H' || whichMatrix == 'h')
+    {
+      h = m;
+      wIsGiven = false;
+      hIsGiven = true;
+    }
+    else
+    {
+      Log::Fatal << "Specify either 'H' or 'W' when creating "
+          "GivenInitialization object!" << std::endl;
+    }
+  }
+
+  // Initialize either H or W, taking control of the given matrix.
+  GivenInitialization(const char whichMatrix, const arma::mat&& m)
```

Comment:
Do you think we should move the `whichMatrix` parameter to the end and set a default value? Also, not sure if we should switch to a boolean here to decide which matrix to use, looks like it would simplify some of the error checks, let me know what you think.

Reply:
Yeah I think it's better to define `whichMatrix` parameter as a boolean here, where `true` means using W matrix and `false` means using H matrix. And this parameter can be set to `true` by default. (initialize W matrix by default)

Reply:
However, I don't think we should move `whichMatrix` to the end as a class variable, since it is only used when the user wants to specify one of W and H matrices. If the user wants to initialize both W and H matrices, then this variable will be useless.